### PR TITLE
Ascii validation

### LINKED ISSUE #103: ASCII validation functions
Hi there - I didn't know if it would make sense to expose ASCII validators as part of the public API. I found this library after finding the https://github.com/lemire/fastvalidate-utf-8 repo, which recommends simdjson, which in turn recommends this. It seems like this library has no (public) functions for validating ASCII.

### PR DESCRIPTION
**Pull request by NicolasJiaxin**
Fixes #103 

This provides the analogous method `validate_ascii` which will check whether a string is ASCII-valid (i.e. the most significant bit of each byte is 0).

### CODE REVIEW

**Comment by NicolasJiaxin** [comment]
@lemire Also, do we need `validate_ascii_brute_force_test` or are the basic tests sufficient?

**Comment by lemire** [comment]
I am reviewing...

File: include/simdutf/implementation.h [hunk 1]
```diff
@@ -220,6 +231,17 @@ class implementation {
    */
   simdutf_warn_unused virtual bool validate_utf8(const char *buf, size_t len) const noexcept = 0;
 
+  /**
+   * Validate the UTF-8 string.
```
Comment (lemire) [inline include/simdutf/implementation.h:235]: Comment should be corrected. :-)

File: include/simdutf/implementation.h [hunk 2]
```diff
@@ -35,6 +35,17 @@ simdutf_really_inline simdutf_warn_unused simdutf::encoding_type autodetect_enco
  */
 simdutf_warn_unused bool validate_utf8(const char *buf, size_t len) noexcept;
 
+/**
+ * Validate the UTF-8 string.
```
Comment (lemire) [inline include/simdutf/implementation.h:39]: Comment should be corrected.

**Comment by lemire** [comment]
I think we can get more clever.

```C++
template<class checker>
 bool generic_validate_ascii(const uint8_t * input, size_t length) {
     buf_block_reader<64> reader(input, length);
     simd::simd8x64<uint8_t> running_or{};
     while (reader.has_full_block()) {
       simd::simd8x64<uint8_t> in(reader.full_block());
       running_or |= in;
       reader.advance();
     }
     uint8_t block[64]{};
     reader.get_remainder(block);
     simd::simd8x64<uint8_t> in(block);
     running_or |= in;
     reader.advance();
     return running_or.is_ascii();
 }
```

The catch with this code is that we are missing the method to do `running_or |= in;`. So we need to implement...

```C++
    simdutf_really_inline simd8x64<T>& operator|=(const simd8<T> &other) { ...
```

in the various kernels...

But the implementation should be something like...


```C++
      this->chunks[0] != other.chunks[0];
      this->chunks[1] != other.chunks[1];
      this->chunks[2] != other.chunks[2];
      this->chunks[3] != other.chunks[3];
     return this;
```

or


```C++
      this->chunks[0] != other.chunks[0];
      this->chunks[1] != other.chunks[1];
     return this;
```

The reason why my proposal is better has to do with the fact that you minimize the number of instructions when you have sizeable inputs. You are basically just unconditionally reading everything and doing a "reduce OR" operation, and then you check on the final result. A "reduce OR" operation is just about as cheap as it gets.

File: src/fallback/implementation.cpp [hunk 3]
```diff
@@ -14,6 +14,10 @@ simdutf_warn_unused bool implementation::validate_utf8(const char *buf, size_t l
     return scalar::utf8::validate(buf, len);
 }
 
+simdutf_warn_unused bool implementation::validate_ascii(const char *buf, size_t len) const noexcept {
+    return scalar::utf8::validate(buf, len);
```
[the remark this answers is not in the document]
Comment (NicolasJiaxin) [reply src/fallback/implementation.cpp:18]: Yes you are right! Good thing you reviewed, I forgot about that...

**Comment by lemire** [comment]
We probably want a fast scalar function to validate ASCII... something like the following (not proven correct, just a sketch !!!).

```
inline simdutf_warn_unused bool validate(const char *buf, size_t len) noexcept {
  const uint8_t *data = reinterpret_cast<const uint8_t *>(buf);
  uint64_t pos = 0;
    // process in blocks of 16 bytes when possible
  for (;pos  + 16 < len; pos += 16) {
      uint64_t v1;
      std::memcpy(&v1, data + pos, sizeof(uint64_t));
      uint64_t v2;
      std::memcpy(&v2, data + pos + sizeof(uint64_t), sizeof(uint64_t));
      uint64_t v{v1 | v2};
      if ((v & 0x8080808080808080) != 0) { return false; }
    }
    // process the tail byte-by-byte
    for (;pos < len; pos ++) {
       if data[pos] >= 0b10000000) { return false; }
    }
    return true;
}
```

File: src/simdutf/ppc64/simd.h [hunk 4]
```diff
@@ -376,6 +376,15 @@ template <typename T> struct simd8x64 {
     this->chunks[3].store(ptr + sizeof(simd8<T>) * 3/sizeof(T));
   }
 
+
```
Comment (lemire) [inline src/simdutf/ppc64/simd.h:379]: The attribute chunks in the simd8x64 struct is marked as const. I think if you just remove the 'const' in 'const chunks', it should work?

Comment (NicolasJiaxin) [reply src/simdutf/ppc64/simd.h:379]: Yes, it works!

Comment (NicolasJiaxin) [reply src/simdutf/ppc64/simd.h:379]: Just not passing the tests yet, but it compiles.

**Comment by NicolasJiaxin** [comment]
@lemire I think it should work now, hopefully.

File: src/fallback/implementation.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (lemire) [reply src/fallback/implementation.cpp]: So the string "é" (or `"\u00e9"`) is not ascii so your tests should return false.

Basically if you take a stream of bytes where all bytes have their most significant bit set to 0, it is ascii, if not then it is not ascii. We want to test both sides of this condition.

**Comment by lemire** [comment]
@NicolasJiaxin  The validation is not perfect, but the validation of the prior code was not perfect, so it is probably good enough.

So I am merging.

We'll need to add the equivalent to AVX-512.

**Pull request merged**